Speed up UART in simulation

cousteaulecommandant opened the pull request:
Simulation UART was set to a speed of 256000 baud, when theoretically it can reach up to almost CLK_FREQ/16. This causes UART to be a bottleneck in apps that print a lot of info, making the simulation take very long to finish.

This commit changes the baud rate to CLK_FREQ/20 (close to the limit), i.e. each UART bit lasts exactly 20 clock cycles.
(In other words, baud rate is now 5,000,000.)
The UART_NCO macro has been rewritten to accommodate this new value without overflowing.

Text-heavy applications have been accelerated about a tenfold.

Note that this does not touch the UART speed in FPGA implementation, which still defaults to 9600 baud (could be accelerated to 115200 but this probably won't be an issue).

davidmallasen commented:
> I see that a new FPGA (ZCU102) was added in the last couple of hours while I was working on this PR, but its x-heep.h doesn't contain the `UART_NCO` macro (probably because the PR is based on an old commit). Should I add it as well?

@LuigiGiuffrida98

LuigiGiuffrida98 commented:
Hi @cousteaulecommandant what is the status of this PR? Can you please rebase it on the main and fix the conflicts?

cousteaulecommandant commented:
@LuigiGiuffrida98 Done!  Turns out the conflict was because someone else had already applied the same fix you requested me to add to ZCU102 (in a slightly different way as I did so it conflicted).

I have solved the conflicts and also changed all the `sw/device/target/*/x-heep.h` files so that they all use the same formula for `UART_NCO`, and added parentheses to the `REFERENCE_CLK_Hz` macro in all of them.

Please, merge this as soon as possible to avoid further conflicts :)

cousteaulecommandant commented:
@LuigiGiuffrida98 @davidmallasen All conflicts solved and all CI tests passed.  The PR is ready to merge! 🎉

davidmallasen commented:
This looks good to me @cousteaulecommandant, thanks! @davideschiavone self-requested a review, so I'll let him check

davideschiavone commented on sw/device/target/sim/x-heep.h:
```diff
@@ -12,15 +12,11 @@ extern "C" {
-#define REFERENCE_CLOCK_Hz 100*1000*1000
-#define UART_BAUDRATE 256000
+#define REFERENCE_CLOCK_Hz (100*1000*1000) // 100 MHz for simulation
+#define UART_BAUDRATE (REFERENCE_CLOCK_Hz / 20) // close to maximum baud rate (/16)
```
true - thx

cousteaulecommandant commented:
Thanks!
